src/operator/random/shuffle_op.cc

```diff
@@ -0,0 +1,132 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * Copyright (c) 2018 by Contributors
+ * \file shuffle_op.cc
+ * \brief Operator to shuffle elements of an NDArray
+ */
+#if (__GNUC__ > 4 && !defined(__clang__major__)) || (__clang_major__ > 4 && __linux__)
+  #define USE_GNU_PARALLEL_SHUFFLE
+#endif
+
+#include <mxnet/operator_util.h>
+#include <algorithm>
+#include <random>
+#include <vector>
+#ifdef USE_GNU_PARALLEL_SHUFFLE
+  #include <parallel/algorithm>
+#endif
+#include "../elemwise_op_common.h"
+
+namespace mxnet {
+namespace op {
+
+namespace {
+
+  template<typename DType, typename Rand>
+  void Shuffle1D(DType* out, index_t size, Rand* prnd) {
+    #ifdef USE_GNU_PARALLEL_SHUFFLE
+      auto rand_n = [prnd](index_t n) {
+        std::uniform_int_distribution<index_t> dist(0, n - 1);
+        return dist(*prnd);
+      };
+      __gnu_parallel::random_shuffle(out, out + size, rand_n);
+    #else
+      std::shuffle(out, out + size, *prnd);
+    #endif
+  }
+
+  template<typename DType, typename Rand>
+  void ShuffleND(DType* out, index_t size, index_t first_axis_len, Rand* prnd) {
+    // Fisher-Yates shuffling
+    const index_t stride = size / first_axis_len;
+    auto rand_n = [prnd](index_t n) {
+      std::uniform_int_distribution<index_t> dist(0, n - 1);
+      return dist(*prnd);
+    };
+    for (index_t i = first_axis_len - 1; i > 0; --i) {
+      index_t j = rand_n(i + 1);
+      if (i != j) {
+        std::swap_ranges(out + stride * i, out + stride * (i + 1), out + stride * j);
+      }
+    }
+  }
+
+}  // namespace
+
+void ShuffleForwardCPU(const nnvm::NodeAttrs& attrs,
+                       const OpContext& ctx,
+                       const std::vector<TBlob>& inputs,
+                       const std::vector<OpReqType>& req,
+                       const std::vector<TBlob>& outputs) {
+  using namespace mxnet_op;
+  if (req[0] == kNullOp) {
+    return;
+  }
+  CHECK_NE(req[0], kAddTo) << "Shuffle does not support AddTo";
+  const TShape input_shape = inputs[0].shape_;
+  const index_t size = inputs[0].Size();
+  const index_t first_axis_len = input_shape[0];
+  Stream<cpu> *s = ctx.get_stream<cpu>();
+  MSHADOW_REAL_TYPE_SWITCH(inputs[0].type_flag_, DType, {
+    Tensor<cpu, 1, DType> in = inputs[0].get_with_shape<cpu, 1, DType>(Shape1(size), s);
+    Tensor<cpu, 1, DType> out = outputs[0].get_with_shape<cpu, 1, DType>(Shape1(size), s);
+    auto& prnd = ctx.requested[0].get_random<cpu, index_t>(ctx.get_stream<cpu>())->GetRndEngine();
+    if (req[0] != kWriteInplace) {
+      std::copy(in.dptr_, in.dptr_ + size, out.dptr_);
+    }
+    if (input_shape.ndim() == 1) {
+      Shuffle1D(out.dptr_, size, &prnd);
+    } else {
+      ShuffleND(out.dptr_, size, first_axis_len, &prnd);
+    }
+  });
+}
+
+
+// No parameter is declared.
+// No backward computation is registered. Shuffling is not differentiable.
+
+NNVM_REGISTER_OP(_shuffle)
```

Comment:
Why register as internal? You can register with name `_random_shuflle` and remove the the python interface.

Reply:
It does not work. `random` module is not generated automatically as [`_OP_NAME_PREFIX_LIST = ['_contrib_', '_linalg_', '_sparse_', '_image_']`](https://github.com/apache/incubator-mxnet/blob/106f97f1881e6bb1a00c56a0ae55200e27297733/python/mxnet/base.py#L455) in `base.py` does not contain `'random'` and `random.py` also does not care about any generated module.

Reply:
Ok, I forgot @piiswrong refactored it. It makes sense to keep the python interface.